Добавление нового функционала спринта 6

Mur-Muring opened the pull request.

Mur-Muring committed:
6 Спринт

Mur-Muring committed:
6 Спринт 1 пытка

Mur-Muring committed:
6 Спринт 1 пытка

Mur-Muring committed:
6 Спринт 1 пытка

Mur-Muring committed:
6 Спринт 1 пытка

Mur-Muring committed:
6 Спринт 1 пытка

Ivanov-s-tmn commented on src/manager/InMemoryHistoryManager.java:
```diff
@@ -1,45 +1,122 @@
+    private Node<Task> tail;
+
+
+    private static class Node<E> {
```
В принципе, нет необходимости в параметризации класса Node, в нашем случае мы всегда храним в качестве данных объекты типа Task и его подклассы, это только немного усложняет синтаксис) А так, молодец, что разобралась и применила параметризацию!

Ivanov-s-tmn commented on src/manager/InMemoryHistoryManager.java:
```diff
@@ -1,45 +1,122 @@
+    private Node<Task> head;
+    private Node<Task> tail;
+
+
```
Лишний отступ

Ivanov-s-tmn commented on src/manager/InMemoryHistoryManager.java:
```diff
@@ -1,45 +1,122 @@
+
+        public Node<E> getNext() {
+            return next;
+        }
```
Геттеры в принципе, тоже не нужны, можно обращаться напрямую к полям ноды `node.task`, `node.prev`, `node.next`, класс вложенный (static private) и нам ничего не грозит)

Ivanov-s-tmn commented on src/manager/InMemoryHistoryManager.java:
```diff
@@ -1,45 +1,122 @@
+                tasksHistory.put(idKey, tail);
+            } else {
+                linkLast(task);
+                tasksHistory.put(idKey, tail);
```
Строки дублируются в обоих случаях, это значит, что они должны выполниться в любом случае.
Достаточно вызова метода `remove()` и `linkLast()`, проверки уже можно вынести в них. В метод `removeNode()`, что пришел не null, а в методе `linkLast()` не нужны дополнительные проверки)

Ivanov-s-tmn commented on src/manager/InMemoryHistoryManager.java:
```diff
@@ -1,45 +1,122 @@
+            next.prev = prev;
+            node.next = null;
+        }
+    }
```
Отлично реализовала алгоритмы добавления ноды, удаления ноды и обхода нод! Сделай методы приватными, они вызываются в публичных методах интерфейса

Ivanov-s-tmn commented on src/manager/InMemoryTaskManager.java:
В методах удаления теперь нужно удалять задачи/подзадачи/эпики из истории

Ivanov-s-tmn commented on test/manager/InMemoryHistoryManagerTest.java:
Нужно покрыть тестами случаи удаления из разных частей истории (из начала/из середины/из конца)
Важно проверить именно порядок задач в истории после удаления

Mur-Muring committed:
6 Спринт 2/6
Добавть удаление из истории просмотров при удадении задач. Реализовать на этой основе Main. Добавть тесты на проверку порядка списка при удалении в истории из начала/середины/конца спика. Улучшить чистоту кода.

Mur-Muring committed:
6 Спринт 2/6
Добавть удаление из истории просмотров при удадении задач. Реализовать на этой основе Main. Добавть тесты на проверку порядка списка при удалении в истории из начала/середины/конца спика. Улучшить чистоту кода.

Mur-Muring committed:
6 Спринт 2/6
Добавть удаление из истории просмотров при удадении задач. Реализовать на этой основе Main. Добавть тесты на проверку порядка списка при удалении в истории из начала/середины/конца спика. Улучшить чистоту кода.

Ivanov-s-tmn commented on src/manager/InMemoryHistoryManager.java:
```diff
@@ -64,11 +55,10 @@ public void add(Task task) {
             } else {
                 linkLast(task);
-                tasksHistory.put(idKey, tail);
             }
```
Достаточно просто вызова методов без проверки на наличие ключа:
```java
remove(idKey);
linkLast(task);
```
если в remove() передать айди несуществующей задачи, то в методе `removeNode()` отработает проверка 
`if (node != null)`
а для linkLast() нам без разницы, мы всегда добавляем ноду в конец)

Mur-Muring replied:
Поняла, исправила

Ivanov-s-tmn commented on src/manager/InMemoryTaskManager.java:
```diff
@@ -1,5 +1,9 @@
 
 package manager;
+/*
+1.Добавила во всех методах, связвнных с удалением по id - удалении из истории, реализовала Main, там все работало,
+но сломало мне тесты по удалению по id
+ */
```
А в методах удаления всех задач/подзадач/эпиков? Они тоже должны удалять из истории)

Ivanov-s-tmn replied:
Этот момент остается актуальным, в методах `deleteTasks()`, `deleteSubtasks()` и `deleteEpics()` нужно добавить логику удаления из истории

Ivanov-s-tmn commented on test/manager/InMemoryTaskManagerTest.java:
```diff
@@ -1,7 +1,6 @@
-- Модернизировала тест deleteSubtaskTest(), теперь он дополнительно проверяет, что id удаленной подзадачи
-не храниться в эпике
-- Добавила тесты влияние set на менеджер
+1. В результате удаления из истории по id, сломались тесты по удаление, вопрос решился путем вызова (get)
+задач и в результате добавления их в историю
```
В каком плане сломались? Просто при удалении задач, если они не были до этого просмотрены методами get, не должно возникать ошибок, так как в менеджере истории есть проверки ноды на null перед удалением задач из истории в методе `removeNode()`, нужно продебажить этот момент, почему возникает и поправить. Если не получится, мне нужна более подробная инфа об ошибке, чтобы помочь)

Mur-Muring replied:
Все работает, моя ошибка в том, что я неправильно чинила тест. проверка в методе removeNode() на null решила все проблемы

Ivanov-s-tmn commented on test/manager/InMemoryHistoryManagerTest.java:
```diff
@@ -93,5 +88,38 @@ public void addTaskTest() {
+
+        int idDelete = task3.getIdTask();
+        historyManager.remove(idDelete);
+        Assertions.assertEquals(historyManager.getHistory().get(idDelete - 1), task2,
+                "Порядок списка нарушен(удаление из середины");
+        Assertions.assertEquals(historyManager.getHistory().get(idDelete), task4,
+                "Порядок списка нарушен(удаление из середины");
```
Тут лучше сравнивать весь список, можно через один ассерт по типу:
```java
Assertions.assertEquals(List.of(task1, task2, task4, task5), historyManager.getHistory(),
                "Порядок списка нарушен(удаление из середины");
```

Mur-Muring replied:
Спасибо, еще на практике не пользовались List.of() и не подумала про него

Mur-Muring committed:
6 Спринт 3/5

Доработать тесты удаления по id, удаления из середины истории. Доработать метод добавления в историю

Ivanov-s-tmn commented on test/manager/InMemoryHistoryManagerTest.java:
```diff
@@ -104,9 +104,7 @@ public void deleteByIdTest() {
-        Assertions.assertEquals(historyManager.getHistory().get(idDelete - 1), task2,
-                "Порядок списка нарушен(удаление из середины");
-        Assertions.assertEquals(historyManager.getHistory().get(idDelete), task4,
+        Assertions.assertEquals(List.of(task1, task2, task4, task5), historyManager.getHistory(),
```
В проверках ниже тоже можно применить это)

Mur-Muring committed:
6 Спринт 4/5

Доработать методы удаления всех подзадач, эпиков, подзадач

Mur-Muring committed:
6 Спринт 4/5

Доработать методы удаления всех подзадач, эпиков, подзадач.
Доработать тесты удаляющие по ID
